File: octo/msgp/msgp-server/src/main/scala/com/sankuai/octo/mworth/dao/worthEvent.scala
package com.sankuai.octo.mworth.dao

import java.util.Date

import com.sankuai.msgp.common.config.DbConnection
import com.sankuai.msgp.common.model.Page
import com.sankuai.octo.mworth.db.Tables._
import org.slf4j.{Logger, LoggerFactory}

import scala.slick.driver.MySQLDriver.simple._
import scala.slick.jdbc.GetResult
import scala.slick.jdbc.StaticQuery.interpolation
import scala.slick.lifted.CanBeQueryCondition

/**
  * Created by zava on 15/11/30.
  */
object worthEvent {
  val LOG: Logger = LoggerFactory.getLogger(worthEvent.getClass)

  private val db = DbConnection.getWorthPool()

  case class WEvent(id: Long, project: String, model: String, functionName: String, functionDesc: String,
                    operationSourceType: Int, business: Int = -1, operationSource: String, targetAppkey: String,
                    appkeyOwt: String, signid: String, startTime: Long, endTime: Long, createTime: Long)

  case class WEventCount(username: String, project: String, model: String, count: Int)

  implicit val getWorthEventCountResult = GetResult(r => WEventCount(r.<<, r.<<, r.<<, r.<<))

  case class WEventBusinessUserOwtModelCount(business: Int, username: String, appkeyOwt: String, project: String, model: String, count: Int)

  implicit val getWorthBusinessUserOwtModelCountResult = GetResult(r => WEventBusinessUserOwtModelCount(r.<<, r.<<, r.<<, r.<<, r.<<, r.<<))

  case class WEventOwtAppkeyModelCount(owt: String, appkey: String, project: String, model: String, count: Int)

  implicit val getWorthOwtAppkeyModelCountResult = GetResult(r => WEventOwtAppkeyModelCount(r.<<, r.<<, r.<<, r.<<, r.<<))

  case class WEventModelMethodCount(project: String, model: String, functionName: String, count: Int)

  implicit val getWorthModelMethodCountResult = GetResult(r => WEventModelMethodCount(r.<<, r.<<, r.<<, r.<<))

  case class WEventDailyCount(username: String, module: String, functionDesc: String, count: Int)

  implicit val getWorthEventDailyCountResult = GetResult(r => WEventDailyCount(r.<<, r.<<, r.<<, r.<<))


  def search(project: Option[String], model: Option[String], functionName: Option[String], targetAppkey: Option[String], fromTime: Date, toTime: Date, page: Page) = {
    db withSession {
      implicit session: Session =>
        val statement = WorthEvent.optionFilter(project)(_.project === _)
          .optionFilter(functionName)(_.functionName === _)
          .optionFilter(targetAppkey)(_.targetAppkey === _)
          .filter(x => x.createTime > fromTime.getTime && x.createTime < toTime.getTime)
        val limit = page.getPageSize
        val offset = page.getStart
        val count = statement.length.run
        page.setTotalCount(count)
        statement.drop(offset).take(limit).list
    }
  }

  def querySign(project: String, functionName: String, signid: Option[String]) = {
    db withSession {
      implicit session: Session =>
        WorthEvent.filter(x => x.project === project
          && x.functionName === functionName)
          .optionFilter(signid)(_.signid === _).sortBy(x => x.id.desc).list
    }
  }


  def insert(wEvent: WEvent): Long = {
    db withSession {
      implicit session: Session =>
        (WorthEvent returning WorthEvent.map(_.id)) +=
          WorthEventRow(0, wEvent.project, wEvent.model, wEvent.functionName, wEvent.functionDesc,
            wEvent.operationSourceType, wEvent.business, wEvent.operationSource,
            Some(wEvent.targetAppkey), Some(wEvent.appkeyOwt), Some(wEvent.signid), Some(wEvent.startTime),
            Some(wEvent.endTime), wEvent.createTime)
    }
  }

  def batchInsert(list: List[WEvent]) = {
    db withSession {
      implicit session: Session =>
        val seq = list.map(wEvent => {
          WorthEventRow(0, wEvent.project, wEvent.model, wEvent.functionName, wEvent.functionDesc,
            wEvent.operationSourceType, wEvent.business, wEvent.operationSource,
            Some(wEvent.targetAppkey), Some(wEvent.appkeyOwt), Some(wEvent.signid), Some(wEvent.startTime),
            Some(wEvent.endTime), wEvent.createTime)
        })
        WorthEvent ++= seq
    }
  }

  /**
    *
    * 统计指定时间段的 用户访问次数
    *
    */
  def countUserModel(start: Long, end: Long) = {
    db withSession {
      implicit session: Session =>
        sql"select operation_source as username,project,  model ,count(1) from worth_event where  create_time  BETWEEN  ${start} and ${end}  and operation_source_type = 0  GROUP BY operation_source ,model".as[WEventCount].list
    }
  }

  def countBusinessUserOwtModel(start: Long, end: Long) = {
    db withSession {
      implicit session: Session =>
        sql"select business, operation_source as username, appkey_owt,project,  model ,count(1) from worth_event where  create_time  BETWEEN  ${start} and ${end}  and operation_source_type = 0  GROUP BY business,username,appkey_owt ,model".as[WEventBusinessUserOwtModelCount].list
    }
  }

  def countOwtAppkeyModel(start: Long, end: Long) = {
    db withSession {
      implicit session: Session =>
        sql"select appkey_owt, target_appkey as appkey , project,  model ,count(1) from worth_event where  create_time  BETWEEN  ${start} and ${end}  and operation_source_type = 0  GROUP BY appkey_owt,appkey,model".as[WEventOwtAppkeyModelCount].list
    }
  }

  def countModelMethod(start: Long, end: Long) = {
    db withSession {
      implicit session: Session =>
        sql"select  project,  model ,function_desc,count(1) from worth_event where  create_time  BETWEEN  ${start} and ${end}  and operation_source_type = 0  GROUP BY model ,function_desc".as[WEventModelMethodCount].list
    }
  }

  /**
    *
    * 统计每一天的 owt-functionDesc
    *
    */
  def countBusinessDescDaily(start: Long, end: Long) = {
    db withSession {
      implicit session: Session =>
        sql"select operation_source as username, model as module, function_desc as functionDesc ,count(1) as count from worth_event where create_time  BETWEEN  ${start} and ${end} and operation_source_type = 0 group by operation_source, function_desc".as[WEventDailyCount].list
    }
  }

  /**
    * 获取没有部门数据的用户信息
    */
  def getBusinessUser() = {
    db withSession {
      implicit session: Session =>
        sql"select  DISTINCT  operation_source from worth_event where business = -1 and operation_source_type = 0;".as[String].list
    }
  }

  def updateBusiness(username: String, business: Int) = {
    db withSession {
      implicit session: Session =>
        sql" update worth_event set business=${business}  where operation_source = ${username} and operation_source_type = 0;".as[Int].first
    }
  }

  def getTargetAppkey() = {
    db withSession {
      implicit session: Session =>
        sql"select  DISTINCT  target_appkey from worth_event where length(target_appkey)>0 and  appkey_owt is NULL;".as[String].list
    }
  }

  def updateAppkeyOwt(appkey: String, owt: String) = {
    db withSession {
      implicit session: Session =>
        sql" update worth_event set appkey_owt=${owt}  where target_appkey = ${appkey}".as[Int].first
    }
  }

  implicit class QueryHelper[T, E](query: Query[T, E, Seq]) {
    def optionFilter[X, R: CanBeQueryCondition](name: Option[X])(f: (T, X) => R) =
      name.map(v => query.withFilter(f(_, v))).getOrElse(query)
  }

}

File: octo/msgp/mworth-common/src/main/java/com/sankuai/octo/mworth/common/model/WorthEvent.java
package com.sankuai.octo.mworth.common.model;



import java.util.Date;
import java.util.List;
import java.util.Map;

/**
 * 价值事件
 */
public class WorthEvent {

    private String project;
    private String model;

    private String functionName;
    private String functionDesc;
    //操作源类型
    private OperationSourceType operationSourceType = OperationSourceType.HUMAN;
    //操作源, 机器/人
    private String operationSource;
    //目标appkey
    private String targetAppkey;

    private int business = -1;


    private String appkeyOwt;
    //签名,主要是用于有开关状态的服务
    private String signid;
    //创建时间
    private Long startTime;
    //结束时间
    private Long endTime;
    //创建时间
    private Date createTime = new Date();


    public WorthEvent(){

    }
    public WorthEvent(String project,String model,String functionDesc,String functionName){
        this.project = project;
        this.model = model;
        this.functionDesc = functionDesc;
        this.functionName = functionName;
    }

    public Date getCreateTime() {
        return createTime;
    }

    public void setCreateTime(Date createTime) {
        this.createTime = createTime;
    }

    public String getProject() {
        return project;
    }

    public void setProject(String project) {
        this.project = project;
    }

    public String getModel() {
        return model;
    }

    public void setModel(String model) {
        this.model = model;
    }

    public String getFunctionDesc() {
        return functionDesc;
    }

    public void setFunctionDesc(String functionDesc) {
        this.functionDesc = functionDesc;
    }

    public String getFunctionName() {
        return functionName;
    }

    public void setFunctionName(String functionName) {
        this.functionName = functionName;
    }

    public OperationSourceType getOperationSourceType() {
        return operationSourceType;
    }

    public void setOperationSourceType(OperationSourceType operationSourceType) {
        this.operationSourceType = operationSourceType;
    }

    public String getOperationSource() {
        return operationSource;
    }

    public void setOperationSource(String operationSource) {
        this.operationSource = operationSource;
    }

    public String getTargetAppkey() {
        return targetAppkey;
    }

    public void setTargetAppkey(String targetAppkey) {
        this.targetAppkey = targetAppkey;
    }

    public String getSignid() {
        return signid;
    }

    public void setSignid(String signid) {
        this.signid = signid;
    }

    public Long getStartTime() {
        return startTime;
    }

    public void setStartTime(Long startTime) {
        this.startTime = startTime;
    }

    public Long getEndTime() {
        return endTime;
    }

    public void setEndTime(Long endTime) {
        this.endTime = endTime;
    }

    public int getBusiness() {
        return business;
    }

    public void setBusiness(int business) {
        this.business = business;
    }
    public String getAppkeyOwt() {
        return appkeyOwt;
    }

    public void setAppkeyOwt(String appkeyOwt) {
        this.appkeyOwt = appkeyOwt;
    }


}